src/containers/Tablets/Tablets.tsx

```diff
@@ -0,0 +1,166 @@
+import {
+    getTabletsInfo,
+    clearWasLoadingFlag,
+    setStateFilter,
+    setTypeFilter,
+} from '../../store/reducers/tablets';
+
+import './Tablets.scss';
+
+const b = cn('tablets');
+
+interface TabletsProps {
+    path?: string;
+    nodeId?: string | number;
+    className?: string;
+}
+
+export const Tablets = ({path, nodeId, className}: TabletsProps) => {
+    const dispatch = useDispatch();
+
+    const {
+        data = {},
+        wasLoaded,
+        loading,
+        error,
+        stateFilter,
+        typeFilter,
+    } = useTypedSelector((state) => state.tablets);
+    const {autorefresh} = useTypedSelector((state) => state.schema);
+
+    const {TabletStateInfo: tablets = []} = data;
+
+    const fetchData = useCallback(
+        (isBackground) => {
+            if (!isBackground) {
+                dispatch(clearWasLoadingFlag());
+            }
+            if (nodeId) {
+                dispatch(getTabletsInfo({nodes: [String(nodeId)]}));
+            } else if (path) {
+                dispatch(getTabletsInfo({path}));
+            }
+        },
+        [path, nodeId, dispatch],
+    );
+
+    useAutofetcher(fetchData, [fetchData], autorefresh);
+
+    const [tabletsToRender, setTabletsToRender] = useState<TTabletStateInfo[]>([]);
+
+    useEffect(() => {
+        let filteredTablets = tablets;
+
+        if (typeFilter.length > 0) {
+            filteredTablets = filteredTablets.filter((tablet) =>
+                typeFilter.some((filter) => tablet.Type === filter),
+            );
+        }
+        if (stateFilter.length > 0) {
+            filteredTablets = filteredTablets.filter((tablet) =>
+                stateFilter.some((filter) => tablet.State === filter),
+            );
+        }
+        setTabletsToRender(filteredTablets);
+    }, [tablets, stateFilter, typeFilter]);
+
+    const handleStateFilterChange = (value: string[]) => {
+        dispatch(setStateFilter(value as ETabletState[]));
+    };
+
+    const handleTypeFilterChange = (value: string[]) => {
+        dispatch(setTypeFilter(value as EType[]));
+    };
+
+    const onShowTooltip = (...args: Parameters<typeof showTooltip>) => {
+        dispatch(showTooltip(...args));
+    };
+
+    const onHideTooltip = () => {
+        dispatch(hideTooltip());
+    };
+
+    const renderTablet = (tabletIndex: number) => {
+        return (
+            <Tablet
+                onMouseLeave={onHideTooltip}
+                onMouseEnter={onShowTooltip}
+                tablet={tabletsToRender[tabletIndex]}
+                key={tabletIndex}
+                className={b('tablet')}
+            />
+        );
+    };
+
+    const renderContent = () => {
+        const states = Array.from(new Set(...[tablets.map((tablet) => tablet.State)]))
+            .filter((state): state is ETabletState => state !== undefined)
+            .map((item) => ({
+                value: item,
+                content: item,
+            }));
```

Comment:
it was 5 traversals: map, spread, set, array.from, map
now it is 6: everything above + filter

I agree that migration from js to ts shouldn't alter the existing code too much, but let's at least remove the spread, it makes no sense